doc/specs/#2046 - Unified keybindings and commands, and synthesized action names.md

```diff
@@ -0,0 +1,589 @@
+    ],
+    "commands": [
+        { "name": "Duplicate Pane", "action": { "action": "splitPane", "split":"auto", "splitMode": "duplicate" }, "icon": null },
+    ]
+}
+```
+
+These two entries are practically the same, except for two key differentiators:
+* the keybinding has a `keys` property, indicating which key chord activates the
+  action.
+* The command has a `name` property, indicating what name to display for the
+  command in the Command Palette.
+
+What if the user didn't have to duplicate this action? What if the user could
+just add this action once, in their `keybindings` or `commands`, and have it
+work both as a keybinding AND a command?
+
+
+## Solution Design
+
+This spec will outline two primary changes to keybindings and commands.
+1. Unify keybindings and commands, so both `keybindings` and `commands` can
+   specify either actions bound to keys, and/or actions bound to entries in the
+   Command Palette.
+2. Propose a mechanism by which actions do not _require_ a `name` to appear in
+   the Command Palette.
+
+
+### Proposal 1: Unify Keybindings and Commands
+
+As noted above, keybindings and commands have nearly the exact same syntax, save
+for a couple properties. To make things easier for the user, I'm proposing
+treating everything in _both_ the `keybindings` _and_ the `commands` arrays as
+**BOTH** a keybinding and a command.
+* Anything entry that has a `keys` value will be added to the keybindings.
+  Pressing that keybinding will activate the action defined in `command`.
+* Anything with a `name`<sup>[1]</sup> will be added as an entry (using that
+  name) to the Command Palette's Action Mode.
+* `keybindings` will be parsed first into keybindings and commands, then
+  `commands` will be layered on top.
+  - The decision to keep both `keybindings` and `commands` as top-level objects
+    in the json blob is certainly **_TODO_: UP FOR DISCUSSION**. Now that we're
+    adding things that might not be bound to keys to the array, `commands` seems
+    like a more appropriate name. However, we're definitely not going to be able
+    to rename `keybindings`.
+
+###### Caveats
+
+* **Nested commands** (commands with other nested commands). If a command has
+  nested commands in the `commands` property, AND a `keys` property, then
+  pressing that keybinding should open the Command Palette directly to that
+  level of nesting of commands.
+* **"Iterable" commands** (with an `iterateOn` property): These are commands
+  that are expanded into one command per profile. These cannot really be bound
+  as keybindings - which action should be bound to the key? They can't all be
+  bound to the same key. If a KeyBinding/Command json blob has a valid
+  `iterateOn` property, then we'll ignore it as a keybinding. This includes any
+  commands that are nested as children of this command - we won't be able to
+  know which of the expanded children will be the one to bind the keys to.

```

Comment:
Future consideration: mapping a key binding to an iterable command opens the palette filtered to the iterable.

Reply:
Well, we can do this for the _nested_ ones, but not the _iterable_ ones

Reply:
Why not?

Reply:
oh oh oh I see, just put the iterable items into the command palette. That's probably doable in the future, I'll make a note of it.